### PR TITLE
Token threshold for posting/commenting/reacting to threads

### PR DESCRIPTION
**Pull request by maxwellfoley3**
## Description
Admins for a token community can now determine a token threshold for the balance a user must have to post. 

This change requires `decimals` information to be added to the database, ie the number of zeroes the token uses to represent. Using `SHOULD_ADD_MISSING_DECIMALS_TO_TOKENS=true`, an updating job should be run in order to port existing entities to the new system.

## Motivation and Context
Details [here](https://www.notion.so/hicommonwealth/Threshold-Token-Forum-2e69c355256f42fb985542c1785dc84b)

## How has this been tested?
Only manually, need to do more systematic QA

## Have proper tags been added (for bug, enhancement, breaking change)?
- [x] yes

## Does this PR affect any server routes?
- [ ] yes, and they are tested: [enter the % coverage here]
- [ ] yes, and they are not tested: [enter the % coverage here]
- [x] yes, but I did not run tests
- [ ] no

### CODE REVIEW

File: client/scripts/helpers/index.ts
Comment (zakhap) [inline client/scripts/helpers/index.ts:354]: What is the data source for this? Why are these methods `expandTokenAmount` and `formatTokenAmount` necessary?

@jnaviask

Comment (jnaviask) [reply client/scripts/helpers/index.ts:354]: It should always be an integer, correct? I don't see what's up with the decimals.

File: client/scripts/views/components/new_thread_form.ts
Comment (zakhap) [inline client/scripts/views/components/new_thread_form.ts:624]: I feel like a lot of these front-end checks for token thresholds should be scoped to only run iff we're in a token forum (which the app should know). We use components like `new_thread_form` in substrate communities as well as cosmos communities, so it probably doesn't make sense to run this kind of filter in the UI unless it's necessary.
Thoughts @jnaviask ?

Comment (zakhap) [reply client/scripts/views/components/new_thread_form.ts:624]: Also, when testing this component, when I didn't have any token in a wallet within the ALEX community (which didn't yet have a threshold set), it through the error `Unable to fetch user's token balance` from the threads controller.

File: client/scripts/views/modals/new_topic_modal.ts
Comment (zakhap) [inline client/scripts/views/modals/new_topic_modal.ts:22]: I feel like `token_threshold` properties should be optional? Like it only matters in Token/ETH-Governance communities

File: client/scripts/views/modals/new_topic_modal.ts
Comment (zakhap) [inline client/scripts/views/modals/new_topic_modal.ts:72]: This shouldn't be visible for Substrate communities. Should check if Token Community

File: client/scripts/views/pages/admin.ts
Comment (zakhap) [inline client/scripts/views/pages/admin.ts:89]: is null allowed for non-eth chains?

File: client/scripts/views/pages/discussions/index.ts
Comment (zakhap) [inline client/scripts/views/pages/discussions/index.ts:70]: Check if token community

File: client/scripts/views/pages/view_proposal/create_comment.ts
Comment (zakhap) [inline client/scripts/views/pages/view_proposal/create_comment.ts:174]: I'm assuming this will also need to check if Token community, bc otherwise it's not necessary to gate user behavior.

File: server/migrations/20210609153902-add-threshold-to-create-topic.js
Comment (zakhap) [inline server/migrations/20210609153902-add-threshold-to-create-topic.js:11]: I feel like we could/should allow Topics to *not* have thresholds. In non-token communities for example.

File: server/routes/createReaction.ts
Comment (zakhap) [inline server/routes/createReaction.ts:46]: first needs to check if token community

**Review by zakhap — CHANGES_REQUESTED** [review]
There are a bunch of the same comments all over the place here, but the "big picture" is that we need to make sure we scope the token threshold UI checks specifically to token communities only and that any route checks happen 1. if token community 2. if token topic threshold set, or otherwise not at all.

File: client/scripts/helpers/index.ts
Comment (maxwellfoley3) [reply client/scripts/helpers/index.ts:354]: I'll make these function names less awkward. We are converting from value denominated in tokens to value denominated in token base units and vice versa. When a user (admin) is typing on the front end how many tokens a threshold should be, it could be something like `0.01`. This gets converted to token base units before being stored in the back end

File: server/migrations/20210609153902-add-threshold-to-create-topic.js
Comment (maxwellfoley3) [reply server/migrations/20210609153902-add-threshold-to-create-topic.js:11]: I was thinking `"0"` would do the job but you're right in a non-token community null makes more sense.

**Comment by maxwellfoley3** [comment]
Ok, I believe I made the necessary changes as well as fixed some bugs and tested it out a bit more.

File: server/routes/createReaction.ts
Comment (maxwellfoley3) [reply server/routes/createReaction.ts:46]: it does on the line prior

**Comment by dillchen** [comment]
@maxwellfoley3 for future reference imo is easier to also mention that each line has been resolved (and mark them as complete) for a review or me whom comes in after the fact

**Comment by dillchen** [comment]
Refreshing on thread creation page.

Steps to recreate:

1. Create a topic threshold that works
2. Create one that doesn't
3. Go to the thread editor
4. Refresh

Screen Shot 2021-07-14 at 3 54 35 AM

and

Screen Shot 2021-07-14 at 3 56 51 AM

**Comment by dillchen** [comment]
@maxwellfoley3 once this is resolved, this is ready to merge

**Comment by maxwellfoley3** [comment]
Okay, the issue @dillchen raised should be resolved

**Comment by zakhap** [comment]
@maxwellfoley3 I can't seem to get this to compile? 

```
ERROR in /.../Documents/Github/cw/eth/types/Erc20DetailedFactory.ts
[tsl] ERROR in /Users/q/Documents/Github/cw/eth/types/Erc20DetailedFactory.ts(5,26)
      TS2307: Cannot find module 'ethers/providers' or its corresponding type declarations.

ERROR in /.../Documents/Github/cw/eth/types/Erc20DetailedFactory.ts
[tsl] ERROR in /Users/q/Documents/Github/cw/eth/types/Erc20DetailedFactory.ts(14,12)
      TS2352: Conversion of type 'Contract' to type 'Erc20Detailed' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'Contract' is missing the following properties from type 'Erc20Detailed': addListener, allowance, approve, balanceOf, and 7 more.
```

**Comment by maxwellfoley3** [comment]
@zakhap Just delete `node_modules` and run `yarn`. Not sure why this is necessary though since I haven't touched any of those files errors appear in relative to master...

**Comment by jnaviask** [comment]
Heads up that this isn't mergeable as things stand, because we're running the chain events listener in a separate PR, so there's no way to communicate across processes to listen to new ERC20s. So we're going to wait on this PR until Timothee's rearchitect is merged, which might deprecate a significant amount of this code as well.

**Comment by zakhap** [comment]
> Heads up that this isn't mergeable as things stand, because we're running the chain events listener in a separate PR, so there's no way to communicate across processes to listen to new ERC20s. So we're going to wait on this PR until Timothee's rearchitect is merged, which might deprecate a significant amount of this code as well.

@dillchen FWIW we'll rebase this against master next week (fix merge conflicts) and then when the chain-events is merged in, we'll get this in too in quick succession.

**Comment by dillchen** [comment]
On review in tokenbalanceCache as well as checks for `chain.type` === `token` add `dao` as valid type

**Comment by jnaviask** [comment]
> On review in tokenbalanceCache as well as checks for `chain.type` === `token` add `dao` as valid type

This won't work because DAOs aren't tokens. DAOs _have_ tokens but we have no logic set up to fetch them.

**Comment by zakhap** [comment]
@maxwellfoley3 Last question before merging in... What do you mean by:
> Using SHOULD_ADD_MISSING_DECIMALS_TO_TOKENS=true, an updating job should be run in order to port existing entities to the new system.

Has this been written? Is this just a migration?

**Pull request merged**
